### PR TITLE
feat: native support in xgrammar for transformers beam search and tests for the LogitsProcessor

### PR DESCRIPTION
**Pull request by SvenPVoigt**
This PR technically introduces tests that weren't present previously and a feature.

## Tests for LogitsProcessor

Previously, there were no tests for the contrib/hf.py LogitsProcessor. Tests have been added in `test_contrib_logits_processor.py`. Here, I create a tiny tokenizer that tokenizes *a* and *b* in a text sequence and a tiny pytorch model that generates the sequence *aab*. The xgrammar is used to restrict the generated sequences to *ab* instead. The test then guarantees that greedy generation generates the sequence of *ab* and that batched generation is supported.

**Tests cover current functionality. Tests also cover new feature.**

## Feature for New LogitsProcessor Functionality

A variant of the xgrammar.contrib.hf.LogitsProcessor that supports beam search in huggingface transformers, as described in #681.

*Example:*

Setup the LogitsProcessor with a parameter indicating it will be stateless (to avoid beam reordering issues).

```diff
-xgr_logits_processor = xgr.contrib.hf.LogitsProcessor(compiled_grammar)
+xgr_logits_processor_beams = xgr.contrib.hf.LogitsProcessor(compiled_grammar, stateless=True)
```

Then pass generate with num_beams.

```diff
-model.generate(**inputs, max_new_tokens=32, logits_processor=[xgr_logits_processor]) #num_beams throws error
+model.generate(**inputs, max_new_tokens=32, logits_processor=[xgr_logits_processor_beams], num_beams=15)
```

#### Motivation

Constrained generation often lacks accuracy compared to plain feature extraction using vanilla llms because constraints distort the probability distributions. However, with vanilla llms, you don't get alignment across documents.

Recent work develops techniques for normalizing the distributions over a single token [Tran-Thien, Fast, High-Fidelity LLM Decoding, 2024][2]. Early work avoids the entire problem of less common tokens by only considering tokenizations of the chosen vocabulary, [De Cao et al., Autoregressive Entity Retrieval, 2021][1].

Alternatively, this PR introduces beam search as a method of improving accuracy by picking the preferred sequence of tokens rather than picking one token at each step.

#### Implementation details

Huggingface already supports beam search, however, it reorders beams. Therefore, the matchers in the current LogitsProcessor cannot track which row they were processing.

The only solution at the moment is to rewalk the input tokens on each generation step to re-obtain beam state.

```diff
class XGRLogitsProcessor(LogitsProcessor):
    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor):
        # Blank mask

        # Refill mask - loop over input_ids rows
            # If terminated, continue
            # New matcher
-           # Consider last token
+           # Rewalk all tokens
            # Fill mask

        # Apply mask to scores
        # Return scores
```

**Current functionality is untouched.** LogitsProcessor works as before when initialized with `stateless=False` *(default)*. The new parameter triggers a different `__call__` to be assigned that rebuilds state.

#### Other Motivation

XGrammar already has a place for huggingface relevant code in xgrammar.contrib.hf - this PR improves the functionality of that code. Therefore, I think this is a natural addition.

[1]: https://arxiv.org/abs/2010.00904
[2]: https://huggingface.co/blog/vivien/llm-decoding-with-regex-constraints

### CODE REVIEW

File: python/xgrammar/contrib/hf.py [hunk 1]
```diff
@@ -112,3 +120,85 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+            self.prompt_len = input_ids.shape[1]
+
+        for i in range(self.batch_size):
+            # New stateless matcher for each row per call
+            matcher = xgr.GrammarMatcher(self.compiled_grammars[i])
+            # Rewalk all input_ids in row to build state
+            for sampled_token in input_ids[i, self.prompt_len :].tolist():
+                assert matcher.accept_token(
+                    sampled_token
+                ), f"Illegal token {sampled_token} allowed in row {i}. Could be that do_sample=True and number of branches in the grammar < num_beams, which forces a -inf scored token into the top num_beams branches. Or it could be a wrong EOS token. Or it could conflict with other logits processors."
```
[the remark this answers is not in the document]
Comment (SvenPVoigt) [reply python/xgrammar/contrib/hf.py]: Did:

```python
ok = matcher.accept_token(sampled_token)
assert ok, "message..."
```

File: python/xgrammar/contrib/hf.py [hunk 2]
```diff
@@ -112,3 +120,85 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+        """
+        Build state from input_ids when ordering of rows is not guaranteed to be consistent, as during beam search.
+
+        After rewalking of the input_ids, fill in bitmask, and apply bitmask to logits.
+
+        Returns:
+            scores: Logits modified with bitmask.
+        """
+        # Lazily initialize GrammarMatchers and bitmask
+        if len(self.matchers) == 0:
```
[the remark this answers is not in the document]
Comment (SvenPVoigt) [reply python/xgrammar/contrib/hf.py]: Did exactly this:

```python
if self.token_bitmask is None:
```

**Comment by SvenPVoigt** [comment]
Should I be reverting to Commit 8bf3651 for the passing unit tests? Unfortunately, I think that is from before when xgrammar switched from nanobind to tvm_ffi.
